### PR TITLE
#3762 Live markdown should scroll rmd and preview

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Similar to VS Code

1. Option to sync text -> preview
2. Option to sync preview -> text
3. Uses markdown line pragmas as in https://github.com/madskristensen/MarkdownEditor/blob/master/src/Margin/BrowserMargin.cs
4. Split code into couple more helper classes

### CODE REVIEW

File: src/Windows/Markdown/Editor/Impl/Preview/Browser/HtmlPageTemplate.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using Microsoft.Common.Core.IO;
+
+namespace Microsoft.Markdown.Editor.Preview.Browser {
+    internal static class HtmlPageTemplate {
+        private static string _htmlTemplate;
+
+        public static string GetPageHtml(IFileSystem fs, string body) {
+            if (_htmlTemplate == null) {
+                var dir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().GetAssemblyPath());
+                var path = Path.Combine(dir, "Markdown", "PreviewTemplate.html");
```
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Browser/HtmlPageTemplate.cs]: Why assembly resource doesn't work?

Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/Preview/Browser/HtmlPageTemplate.cs]: I am following what Mads' editor does. I think technically the file may be editable by the user

Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/Preview/Browser/HtmlPageTemplate.cs]: RS has then as plain files too

File: src/Windows/Markdown/Editor/Impl/Preview/Margin/MarginControls.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            } else {
+                CreateRightMarginControl(parent);
+            }
+        }
+
+        private void CreateRightMarginControl(Decorator parent) {
+            var width = _settings.PreviewWidth;
+
+            var grid = new Grid();
+            grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(0, GridUnitType.Star) });
```
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Margin/MarginControls.cs]:
```cs
var grid = new Grid {
    ColumnDefinitions = {
        new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) },
        new ColumnDefinition { Width = new GridLength(5, GridUnitType.Pixel) },
        new ColumnDefinition { Width = new GridLength(width, GridUnitType.Pixel), MinWidth = 150 }
    },
    RowDefinitions = { new RowDefinition() },
    Children = {
        splitter.SetGridPosition(0, 1),
        _webBrowser.SetGridPosition(0, 2)
    }
}
```

File: src/Windows/Markdown/Editor/Impl/Preview/Browser/HtmlPageTemplate.cs (hunk 1 above)
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Browser/HtmlPageTemplate.cs]: Ok

File: src/Windows/Markdown/Editor/Impl/Preview/Margin/MarginControls.cs (hunk 2 above)
Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/Preview/Margin/MarginControls.cs]: OK

File: src/Windows/Markdown/Editor/Impl/Preview/Parser/Extensions.cs [hunk 3]
```diff
@@ -7,7 +7,7 @@
 namespace Microsoft.Markdown.Editor.Preview.Parser {
     internal static class Extensions {
         public static string GetText(this LeafBlock block) {
-            if(block?.Lines == null) {
+            if(block?.Lines == null || block.Lines.Lines == null) {
```
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Parser/Extensions.cs]: `block?.Lines?.Lines == null`?

File: src/Windows/Markdown/Editor/Impl/Preview/Margin/PreviewMargin.cs [hunk 4]
```diff
@@ -92,13 +115,25 @@ public sealed class PreviewMargin : Border, IWpfTextViewMargin, IMarkdownPreview
             _textChanged = true;
         }
 
+        private void UpdateBrowserScrollPosition() {
+            if(!_settings.ScrollPreviewWithEditor) {
+                return;
+            }
+
+            IdleTimeAction.Create(() => {
+                var lineNumber = _textView.TextSnapshot.GetLineNumberFromPosition(_textView.Caret.Position.BufferPosition);
```
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Margin/PreviewMargin.cs]: Does it matter if this action is called before or after `OnIdle`

File: src/Windows/Markdown/Editor/Impl/Preview/Code/EvalSession.cs [hunk 5]
```diff
@@ -43,7 +43,7 @@ internal sealed class EvalSession : IDisposable {
 
             if (!RSession.IsHostRunning) {
                 var settings = _services.GetService<IRSettings>();
-                await RSession.EnsureHostStartedAsync(new RHostStartupInfo(), SessionCallback, 3000, ct);
+                await RSession.EnsureHostStartedAsync(new RHostStartupInfo(isInteractive: true), SessionCallback, 3000, ct);
```
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Code/EvalSession.cs]: @karthiknadig , @int19h , do you remember what are the side effects of making session interactive?

Comment (karthiknadig) [inline src/Windows/Markdown/Editor/Impl/Preview/Code/EvalSession.cs]: Popup UI is disabled in most libraries when the `Interactive` flag is false.

File: src/Windows/Markdown/Editor/Impl/Preview/Browser/BrowserWindow.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            if (zoomFactor == 100) {
+                return;
+            }
+
+            dynamic OLECMDEXECOPT_DODEFAULT = 0;
+            dynamic OLECMDID_OPTICAL_ZOOM = 63;
+            var fiComWebBrowser = typeof(WebBrowser).GetField("_axIWebBrowser2", BindingFlags.Instance | BindingFlags.NonPublic);
+
+            var objComWebBrowser = fiComWebBrowser?.GetValue(_webBrowser);
+            objComWebBrowser?.GetType().InvokeMember("ExecWB", BindingFlags.InvokeMethod, null, objComWebBrowser, new object[] {
```
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Browser/BrowserWindow.cs]: Don't we have `IWebBrowser2` COM import?

File: src/Windows/Markdown/Editor/Impl/Document/MdEditorDocument.cs [hunk 7]
```diff
@@ -53,12 +54,14 @@ public sealed class MdEditorDocument : IMdEditorDocument {
...
         public IContainedLanguageHandler ContainedLanguageHandler { get; }
+
+        public IContainedLanguageHost ContainedLanguageHost { get; }
         #endregion
 
         #region IDisposable
         public void Dispose() {
             EditorBuffer.RemoveService(this);
-            _projectionBufferManager.Dispose();
+            _projectionBufferManager?.Dispose();
```
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Document/MdEditorDocument.cs]: How is it possible that `_projectionBufferManager` is null?

File: src/Windows/Markdown/Editor/Impl/ContainedLanguage/RExpressionTermFilter.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    /// <summary>
+    /// Helps R parser to handle R Markdown-specific constructs
+    /// </summary>
+    internal sealed class RExpressionTermFilter: IExpressionTermFilter {
+        private readonly ITextBuffer _textBuffer;
+
+        public RExpressionTermFilter(ITextBuffer textBuffer) {
+            Check.ArgumentNull(nameof(textBuffer), textBuffer);
+            _textBuffer = textBuffer; // R language buffer
+            _textBuffer.AddService(this);
```
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/ContainedLanguage/RExpressionTermFilter.cs]: How it is removed from `ITextBuffer`?

Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/ContainedLanguage/RExpressionTermFilter.cs]: It doesn't need to be removed. Generally removing services is optional unless they are disposable or complicated things that can introduce leaks such as when attached to VS COM events.

File: src/Windows/Markdown/Editor/Impl/Document/MdEditorDocument.cs (hunk 7 above)
Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/Document/MdEditorDocument.cs]: Prob never

File: src/Windows/Markdown/Editor/Impl/Preview/Browser/BrowserWindow.cs (hunk 6 above)
Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/Preview/Browser/BrowserWindow.cs]: Maybe... it is borrowed code. Let me see.

Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/Preview/Browser/BrowserWindow.cs]: Its huge http://pinvoke.net/default.aspx/Interfaces/IWebBrowser2.html - I guess Mads didn't want to import the whole thing

File: src/Windows/Markdown/Editor/Impl/Preview/Code/EvalSession.cs (hunk 5 above)
Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/Preview/Code/EvalSession.cs]: It has to be interactive so it behaves like REPL and outputs errors without terminating the host.

File: src/Windows/Markdown/Editor/Impl/Preview/Margin/PreviewMargin.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/Preview/Margin/PreviewMargin.cs]: Generally no. OnIdle only runs when browser pane is active while UpdateBrowserScrollPosition runs when caret position changes which means editor pane has focus.

File: src/Windows/Markdown/Editor/Impl/Preview/Parser/Extensions.cs (hunk 3 above)
Comment (MikhailArkhipov) [inline src/Windows/Markdown/Editor/Impl/Preview/Parser/Extensions.cs]: I tried, it gives me 

```
Error	CS0023	Operator '?' cannot be applied to operand of type 'StringLineGroup'
```

Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Parser/Extensions.cs]: It means that `StringLineGroup` is a struct, so `block?.Lines.Lines == null` is enough

File: src/Windows/Markdown/Editor/Impl/Preview/Margin/PreviewMargin.cs (hunk 4 above)
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Margin/PreviewMargin.cs]: Ok

File: src/Windows/Markdown/Editor/Impl/Preview/Code/EvalSession.cs (hunk 5 above)
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Code/EvalSession.cs]: Ok, just to be sure.

File: src/Windows/Markdown/Editor/Impl/Preview/Browser/BrowserWindow.cs (hunk 6 above)
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/Preview/Browser/BrowserWindow.cs]: Ok

File: src/Windows/Markdown/Editor/Impl/ContainedLanguage/RExpressionTermFilter.cs (hunk 8 above)
Comment (AlexanderSher) [inline src/Windows/Markdown/Editor/Impl/ContainedLanguage/RExpressionTermFilter.cs]: Ok

**Comment by AlexanderSher** [comment]
👍

**Pull request merged**